src/protocol/sort/generate_permutation.rs

```diff
@@ -44,53 +45,23 @@ pub struct ShuffledPermutationWrapper<T, C: Context> {
 /// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
 /// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
 /// <https://eprint.iacr.org/2019/695.pdf>.
-pub(super) async fn shuffle_and_reveal_permutation<
-    F: Field,
-    S: SecretSharing<F> + Reshare<C, RecordId> + Reveal<C, RecordId, Output = F>,
-    C: Context,
->(
-    ctx: C,
-    input_permutation: Vec<S>,
-) -> Result<RevealedAndRandomPermutations, Error> {
-    let random_permutations_for_shuffle = get_two_of_three_random_permutations(
-        u32::try_from(input_permutation.len()).expect("Input size fits into u32"),
-        ctx.narrow(&GeneratePermutation).prss_rng(),
-    );
-
-    let shuffled_permutation = shuffle_shares(
-        input_permutation,
-        (
-            random_permutations_for_shuffle.0.as_slice(),
-            random_permutations_for_shuffle.1.as_slice(),
-        ),
-        ctx.narrow(&ShufflePermutation),
-    )
-    .await?;
-
-    let reveal_ctx = ctx.narrow(&RevealPermutation);
-    let wrapper = ShuffledPermutationWrapper {
-        perm: shuffled_permutation,
-        ctx,
-    };
-    let revealed_permutation = wrapper.reveal(reveal_ctx, NoRecord).await?;
-
-    Ok(RevealedAndRandomPermutations {
-        revealed: revealed_permutation,
-        randoms_for_shuffle: random_permutations_for_shuffle,
-    })
-}
-
-/// This is a malicious implementation of shuffle and reveal.
 ///
 /// Steps
 /// 1. Get random permutation 2/3 shared across helpers
 /// 2. Shuffle shares three times
 /// 3. Validate the accumulated macs - this returns the revealed permutation
-pub(super) async fn malicious_shuffle_and_reveal_permutation<F: Field + ExtendableField>(
-    m_ctx: MaliciousContext<'_, F>,
-    input_permutation: Vec<MaliciousReplicated<F>>,
-    malicious_validator: MaliciousValidator<'_, F>,
-) -> Result<RevealedAndRandomPermutations, Error> {
+pub(super) async fn shuffle_and_reveal_permutation<C, S, F>(
+    m_ctx: C::UpgradedContext<F>,
+    input_permutation: Vec<S>,
+    malicious_validator: C::Validator<F>,
```

Comment:
I remember we had an idea of adding/using `ctx.reveal(shares)` that validates before revealing in malicious setting and just reveals shares one by one in semi-honest.

That requires validator to be inside `UpgradedContext` and a runtime check (or another generic argument) that validator is called only once, but does not require carrying the validator along with the context.

Did you decide that this is not the best way to go?

Reply:
I decided not to make that change here;  I'm still not happy with the way that validators integrate into the process, but maybe we can move it into the upgraded context.  That might be a good follow-up.